### DISCUSSION TITLE
Performance issues with high memory consumption and CPU load for medium-to-large stacks

### DISCUSSION THREAD

**Discussion by hcharley** [pulumi/pulumi#9477] [Pulumi in Practice] [answered]
## Description of topic

I am dealing with RAM/CPU (memory/load) issues with my Node.JS / Typescript Automation API stack when it grows beyond a certain size, and this discussion is for those having similar issues with memory & CPU for any Pulumi user (not just those with my stack).

I am not creating an issue, because I don't have a ready-made reproduction, nor do I have code samples. Instead, I would like to begin to collate various issues/PRs that I've been reading that all touch on this issue. 

## Observed output 

According to PM2, as soon as my Pulumi "program" is finished running (for a preview), there is a consumption of greater than 100% (~135%-145%) CPU and with an available memory of 24GB, it quickly goes up to 14GB and above. I used node options to increase the max-old-space-size to 75% of my machine's total RAM (32 GB). 

PM2 logged:

```
terminate called after throwing an instance of 'std::bad_alloc'
std::bad_alloc
exited with code [0] via signal [SIGABRT]
```

When pausing the script shortly after a hang, the `serializeProperty` seems to be running according to the heap profile:

```
node_modules/.pnpm/@pulumi+pulumi@3.31.0-alpha.1650929512/node_modules/@pulumi/pulumi/runtime/rpc.js
```

### Modified behavior

I have tried selectively targeting single URNs, but even that still seems to trigger the issue.

## Environment

I was using the latest Pulumi version, but I just switched the latest alpha to see if that would help. Here are my versions now: 

```
    "@pulumi/gcp": "^6.20.0",
    "@pulumi/kubernetes": "^3.18.3",
    "@pulumi/postgresql": "^3.4.0",
    "@pulumi/pulumi": "^3.31.0-alpha.1650929512",
    "@pulumi/random": "^4.5.0",

pnpm 6.32.10

pulumi about
CLI          
Version      3.31.0-alpha.1650929512
Go Version   go1.17.9
Go Compiler  gc

Host     
OS       debian
Version  11.3
Arch     x86_64

Backend        
Name           pulumi.com
URL            https://app.pulumi.com/abc
User           abc
Organizations  x, y, z

Ran inside a Github Codespace with 32 GB RAM, 16 cores, 128 storage
```

## Hunches

- My best guess is that some "serialization" or "unmarshaling" process is going wild due to lots of resource dependencies. 
- An issue with the RPC could be causing this issue, expanding the impact beyond just the NodeJS Automation API, to other languages such as Go, Python, etc.
- I have recorded log points showing that parent URNs can't be looked up
- It could be promise leaks or infinite promises being created
- I am thinking that the Kubernetes YAML directory output could help save me since I have a lot of Kubernetes resources, and each of them have a lot of metadata associated with them
- Could some of the metadata associated with Kubernetes resources be removed (like [last-applied-configuration](https://github.com/pulumi/pulumi-kubernetes/issues/1118))?
  - See https://github.com/pulumi/pulumi/issues/8677#issuecomment-1031777766
- Could it be that high network transfer (uploads specifically) are causing a compounding/confounding issue leading to high memory and CPU consumption?
  - See https://github.com/pulumi/pulumi/issues/9190#issuecomment-1067764698
  - #6074
  - #5469

## Related issues & pull requests

Newest first:

- #9472*
- #9467*
- #9457
- #9386
- #9350
- #9348
- https://github.com/pulumi/pulumi-kubernetes-operator/issues/262
- #9168
- #9119
- #9106
- #9049*
- #8872 
- #8950
- #8709 
- #8677
- #8415*
- #8199*
- #7682 
- #7861
- #7958
- #7346*
- https://github.com/pulumi/pulumi-kubernetes/issues/1568
- #6197
- #6074*
- #5469*
- #5001
- #4876*
- #4605*
- #3987
- #2514
- https://github.com/pulumi/pulumi/issues/2484#issue-414791454*
- https://github.com/pulumi/pulumi/pull/1857#issuecomment-677694535*
- #215

\* Not strictly specific to performance issues, but collating here for awareness.

### Searches

- https://github.com/pulumi/pulumi/labels/impact%2Fperformance
- https://github.com/pulumi/pulumi/labels/impact%2Fnon-specific
- [Duplicate issues](https://github.com/pulumi/pulumi/issues?q=label%3Aresolution%2Fduplicate+is%3Aclosed)

**Comment by hcharley** [comment, marked as answer]
I think I may have found my issue. 

Circular serialization of objects passed into a ComponentResource, despite not being included in the output or args, could be causing infinite serialization.

**Reply by hcharley** [reply]
I am going to try and use `doNotCapture` to suppress some of this data from serialization.

**Reply by hcharley** [reply]
I do not know how to pass a custom `serialize` function. And `doNotCapture` does not appear to work how I thought it would.

**Reply by hcharley** [reply]
Okay, so by pre-serializing an object I had been using to construct the `containers` field in a k8 deployment into a JSON string and back into a JS object, I was able to avoid infinite serialization.

TIL (and should have always been doing, as its clean!): always give your pulumi component resources a simple DTO to consume, never a method that uses a JS class instance ~~that returns an array~~.

**Comment by XBeg9** [comment]
Team, in the latest releases of pulumi something really wrong 3.34+, getting `Killed` or OOM easily on a stack that was working fine before.

Thanks

**Reply by rshade** [reply]
are you on 3.35.3? Can you try this, we found an alias issue causing a lot of memory usage?

**Reply by vtereshyn** [reply]
I am using pulumi `3.48.0` and node `18.12.0`. Also, I am using the latest version of Typescript with EcmaScript modules, and issue with OOM has become pretty common. @rshade

**Reply by rshade** [reply]
@vtereshyn can you open an issue on the pulumi repo with some details so we can try to recreate?

**Reply by vtereshyn** [reply]
@rshade done – https://github.com/pulumi/pulumi/issues/11476
